pkg/cloudsources/paladin/client.go

```diff
@@ -161,3 +202,74 @@ func getErrorResponse(resp *http.Response) error {
 	}
 	return fmt.Errorf("request failed with status %d and response %q", resp.StatusCode, buf.String())
 }
+
+func (c *paladinClient) mapAssetToDiscoveredCluster(asset Asset) (*discoveredclusters.DiscoveredCluster, error) {
+	d := &discoveredclusters.DiscoveredCluster{
+		ID:            clusterIDFromAsset(asset),
+		Name:          asset.Name,
+		Type:          clusterTypeFromAsset(asset),
+		ProviderType:  providerTypeFromAsset(asset),
+		Region:        asset.Region,
+		CloudSourceID: c.cloudSourceID,
+	}
+
+	firstDiscoveredAt, err := gogoProto.TimestampProto(asset.FirstDiscoveredAt)
+	if err != nil {
+		return nil, errox.InvariantViolation.New("converting timestamps")
+	}
+	d.FirstDiscoveredAt = firstDiscoveredAt
+
+	return d, nil
+}
+
+func providerTypeFromAsset(a Asset) storage.DiscoveredCluster_Metadata_ProviderType {
+	switch a.Source {
+	case "gcp":
+		return storage.DiscoveredCluster_Metadata_PROVIDER_TYPE_GCP
+	case "aws":
+		return storage.DiscoveredCluster_Metadata_PROVIDER_TYPE_AWS
+	case "azure":
+		return storage.DiscoveredCluster_Metadata_PROVIDER_TYPE_AZURE
+	default:
+		return storage.DiscoveredCluster_Metadata_PROVIDER_TYPE_UNSPECIFIED
+	}
+}
+
+func clusterTypeFromAsset(a Asset) storage.ClusterMetadata_Type {
+	switch a.Type {
+	case clusterTypeAKS:
+		return storage.ClusterMetadata_AKS
+	case clusterTypeGKE:
+		return storage.ClusterMetadata_GKE
+	case clusterTypeEKS:
+		return storage.ClusterMetadata_EKS
+	default:
+		return storage.ClusterMetadata_UNSPECIFIED
+	}
+}
+
+func clusterIDFromAsset(a Asset) string {
+	// If the type is not AKS, we can safely return the ID from Paladin Cloud.
+	// In the case of EKS, it will be the ARN, in the case of GCP it will be the cluster ID.
+	if a.Type != clusterTypeAKS {
+		return a.ID
+	}
+
+	// For Azure, our cluster ID is different from the one specified by Paladin Cloud.
+	// Paladin CLoud has the Azure reference, which is of the format
+	//	"subscriptions/<id>/resourcegroups/<group>/azure resource names.../<cluster name>"
+	// Internally, our format for AKS clusters is
+	//  "MC_<resource group>_<cluster name>_<location>
+
+	// Index 3 -> resource group name
+	// Index 7 -> cluster name
+	idParts := strings.Split(a.ID, "/")
```

Comment:
Alternatively to parsing the ID, we could also extend the asset fields we fetch from PaladinCloud. They return subscription id, resource group, cluster name and location as separate fields.

Reply:
I'd rather refrain from it, the resource group is currently under `account name` and I'm not sure this is correct (in Azure, resource group != account name). So, I'd rather parse the ID for the time being, given that this is less likely to change.